lib/experimental/auto-inserting-blocks.php

```diff
@@ -0,0 +1,246 @@
+<?php
+/**
+ * Auto-inserting blocks.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Return a function that auto-inserts blocks relative to a given block.
+ *
+ * @param array  $inserted_block    The block to insert.
+ * @param string $relative_position The position relative to the given block.
+ * @param string $anchor_block      The block to insert relative to.
+ * @return callable A function that accepts a block's content and returns the content with the inserted block.
+ */
+function gutenberg_auto_insert_block( $inserted_block, $relative_position, $anchor_block ) {
+	return function( $block ) use ( $inserted_block, $relative_position, $anchor_block ) {
+		if ( $anchor_block === $block['blockName'] ) {
+			if ( 'first_child' === $relative_position ) {
+				array_unshift( $block['innerBlocks'], $inserted_block );
+				// Since WP_Block::render() iterates over `inner_content` (rather than `inner_blocks`)
+				// when rendering blocks, we also need to prepend a value (`null`, to mark a block
+				// location) to that array.
+				array_unshift( $block['innerContent'], null );
+			} elseif ( 'last_child' === $relative_position ) {
+				array_push( $block['innerBlocks'], $inserted_block );
+				// Since WP_Block::render() iterates over `inner_content` (rather than `inner_blocks`)
+				// when rendering blocks, we also need to prepend a value (`null`, to mark a block
+				// location) to that array.
+				array_push( $block['innerContent'], null );
+			}
+			return $block;
+		}
+
+		$anchor_block_index = array_search( $anchor_block, array_column( $block['innerBlocks'], 'blockName' ), true );
+		if ( false !== $anchor_block_index && ( 'after' === $relative_position || 'before' === $relative_position ) ) {
+			if ( 'after' === $relative_position ) {
+				$anchor_block_index++;
+			}
+			array_splice( $block['innerBlocks'], $anchor_block_index, 0, array( $inserted_block ) );
+
+			// Find matching `innerContent` chunk index.
+			$chunk_index = 0;
+			while ( $anchor_block_index > 0 ) {
+				if ( ! is_string( $block['innerContent'][ $chunk_index ] ) ) {
+					$anchor_block_index--;
+				}
+				$chunk_index++;
+			}
+			// Since WP_Block::render() iterates over `inner_content` (rather than `inner_blocks`)
+			// when rendering blocks, we also need to insert a value (`null`, to mark a block
+			// location) into that array.
+			array_splice( $block['innerContent'], $chunk_index, 0, array( null ) );
+		}
+		return $block;
+	};
+}
```

Comment:
how often do we know in advance the block we want to insert and then need to create a function to insert that? what if we instead made a function that actually inserts the block and skipped the anonymous function creation?

if we know we want to always insert the same block we can create our callback with a static value or an enclosed value.

also is this function different than hooking into the existing filters? I wonder if there's reason to insert after more than one block type, in which case we start duplicating code or adding new interfaces, but we already have the ability to do this if we use something like `render_block` with an appropriate callback.

Reply:
> how often do we know in advance the block we want to insert and then need to create a function to insert that? what if we instead made a function that actually inserts the block and skipped the anonymous function creation?
>
> if we know we want to always insert the same block we can create our callback with a static value or an enclosed value.

Not sure I'm reading you correctly, but the point of this API is to allow 3rd parties to have their blocks auto-inserted next to pretty much any other block, so we can't really know in advance what blocks will need to be inserted.

> also is this function different than hooking into the existing filters? I wonder if there's reason to insert after more than one block type, in which case we start duplicating code or adding new interfaces, but we already have the ability to do this if we use something like `render_block` with an appropriate callback.

It's definitely possible to insert a block after (or next to) more than one block type, e.g. a like button after Post Content, or as Comment Template's last child.

An earlier version of this (https://github.com/WordPress/gutenberg/pull/51294) used `render_block`, but that had a number of drawbacks -- most notably, it was hard to find out whether or not the block was rendered as part of a user-modified or unmodified template. In the present version of the code, we get this information pretty much for free from the `get_block_templates` and `get_block_file_template` filters that we've using.
Furthermore, while the `render_block` filter worked fine for `before`/`after` insertion, it's less practical for `first_child` or `last_child` insertion, which inevitably requires some parsed tree structure of blocks; in #51294, I was using `render_block_data` to that effect.

That aside, I'm not sure how a callback for `render_block` would look substantially different from what we're doing here that would allow it to eschew the indirection? The problem we're facing is binding, isn't it? We'd like our API to allow people to specify what block they want to auto-insert next to what other block; but pretty much all existing block filters have at best one param that communicates the block that is currently being parsed/rendered/whatever (the anchor block, in our nomenclature), so in order to pass along the "other", auto-inserted block, we need some mechanism to accommodate for that. In Core, I'd do that via a dedicated registry for auto-inserted blocks ([as noted here](https://github.com/WordPress/gutenberg/pull/51449/files#diff-b6a5f5374aa2b01e4369049937355b47072d9123e8918523bc49b46efa501807R121)), from which the filters could then read which block needs to be inserted next to the currently processed anchor block; but for the sake of something more self-contained that can be easily implemented within Gutenberg, the filter factories seemed like a fair compromise.

Reply:
> inserted next to pretty much any other block, so we can't really know in advance what blocks will need to be inserted.

this was kind of what led me to ask in the first place, because this function signature implies that we have to know in advance the block we want to insert plus the block type we want to anchor it to.

> so in order to pass along the "other", auto-inserted block, we need some mechanism to accommodate for that

I wasn't commenting on the mechanism to do this, but the interface we present, which immediately creates an anonymous function and hides the ability to make decisions about insertion based on the current block.

it seems like if we want to auto-insert after two block types that we have to call `gutenberg_auto_insert_block` twice, one for each block type. it seems like if we only want to auto-insert based on some block attribute then there's no way to do that.

I'm wondering if we were to invert this so that the consumer passed in the logic for where to insert if it could be less constrained and less complicated. we could pass in an anonymous function which returns a relative position and block to insert, if one ought to be inserted, or a filter that does the same.

```php
add_filter( 'block_auto_insert', function ( $block ) {
	if ( $block['attrs']['show_thing'] ) {
		return array( 'where' => 'after', 'block' => array( … );
	}
} );
```

```php
add_filter( 'gutenberg_serialize_block', function ( $block ) {
	$auto_insert = apply_filter( 'block_auto_insert', $block );

	if ( null === $auto_insert ) {
		return $block;
	}

	list( 'where' => $where, 'block' => $inserted_block ) = $auto_insert;
	switch ( $where ) {
		case 'before':
			…

		…
	}
} );
```